### ISSUE TITLE
LogBookViewer causing slow-down.

### ISSUE THREAD

**Issue by melvinzhang** [magarena/magarena#419]
Submitter: Anonymous
LogBookViewer progressively slows down game

hey the LogBookViewer is causing serious slowdown. gets worse the longer a duel gets. toggling log on and off in ui has no effect. so i commented most of the LogBookViewer.java code out and rebuilt and hey viola, game is running perfectly even in very long duels. might wanna have that looked at in more depth.

**Comment by lodici** [comment]
Thanks - will take a look.

**Comment by melvinzhang** [comment]
confirmed by manually adding 1000 log messages first before starting the game. Some preliminary investigation suggest the issue may be with LogStackViewer.setLogStackLayout which is called on every update, the removal and adding back of the two viewers is quite slow. Perhaps the constraints can be specified in the layout manage once in the constructor, instead of modified on every update.

**Comment by lodici** [comment]
I think it is more to do with the fact that LogBookViewer.update() and LogBookViewer.addMessagePanel() create a MessagePanel for every message in MagicLogBook even though you can only ever see the last X messages in the on-screen log. I should have a fix that limits the number of MessagePanels to just those that are visible sometime today.

**Comment by melvinzhang** [comment]
That's what I thought at first, and I tested by putting a debug statement to print whenever update is called. It turns out not to be the case, as the gameplay is slow even when update is not called.  

Tracing thought the code, the reason is that LogBookViewer.update is only called during an undo when message are removed from the MagicLogBook via MagicLogBook.removeTo. It does not affect the regular play of the game.

However, having a large log with many items, slow down regular play that does not have any undo, i.e. without any calls to LogBookViewer.update().

**Comment by melvinzhang** [comment]
I tested that removing setLogStackLayout(), and putting the layout once in the constructor removes the gameplay slowdown due to a big log. However, I was unable to replicate the current behavior of the stack/log viewer, as I could not figure out the proper layout hints to get the same functionality as setLogStackLayout

Limiting the number of message panels in the log to only the visible ones sounds like a promising approach as well.

**Comment by melvinzhang** [comment]
Added the test case I was using, steps to replicate given in comments in the test case, replicated here for convenience

Steps for TestSlowLog
1. Click on Lighting Bolt
2. Click on Undo
3. wait for 1000 log messages to appear in log
4. Try to continue the game normally, it will be slower than before

**Comment by lodici** [comment]
setLogStackLayout() is affected but is not the cause. If you have large number of MessagePanel components then it is going to impact on layout (even though you can only see the last 20 or so). Trying switching between full screen mode and you should see significant delays in screen refresh. I have a fix which seems to alleviate the problem (still testing) - I doctored MagicLogBook.add() to create 100 copies of each message then ran AI vs AI - MagicLogBook size reached 7000+ with no discernible slow-down so far.

**Comment by melvinzhang** [comment]
Agree, changing only setLogStackLayout has no impact on update (called during undo) or window resize. Limiting the number of MessagePanel created is a more general fix and as a bonus reduces memory usage.

**Comment by lodici** [comment]
I added the updates to restrict the number of referenced MessagePanels to only those that are visible in the on-screen viewer. Also undo only looks at the last X messages that will be visible which should remove the delay demonstrated in TestSlowLog when there are a lot of log messages. I have left this issue open since the fix has introduced an edge case regarding a missing on-screen log message when using a mulligan but I think the performance issue is too important for that to prevent the commit.

**Comment by melvinzhang** [comment]
What are the steps to reproduce the edge case?

Is the log viewer suppose to be scrollable? I am unable to scroll back in the viewer, even before the commit.

**Comment by lodici** [comment]
> Is the log viewer suppose to be scrollable? 

No. 

> What are the steps to reproduce the edge case?

Start a duel and at earliest opportunity return back to main menu. Start next duel. Mulligan down to three cards. It does not display your mulligan steps in the log. It is related to layout - for some reason the scrollpane is not sized correctly in this instance.

**Comment by melvinzhang** [comment]
I'm unable to replicate the edge case. 

My steps were:
1. New duel
2. Next ->
3. Game 1 ->
4. Play this hand ->
5. Hit esc, select quit to main menu
6. New duel
7. Next ->
8. Game 1 ->
9. Mulligan down to 3 cards
10. Play this hand ->

Log shows two entries with Player and AI mulligan choices.

**Comment by lodici** [comment]
That is correct - I am still able to replicate at will. Strange.

**Comment by melvinzhang** [comment]
tracking the mulligan problem in a separate issue, closing this one as the performance problem is solved.
